src/Motor.Extensions.Hosting.Kafka/KafkaMessageConsumer.cs

```diff
@@ -180,61 +197,167 @@ private async Task SingleMessageHandlingAsync(ConsumeResult<string?, byte[]> msg
+
+        return new ConsumeResultAndProcessedMessageStatus(msg, ProcessedMessageStatus.CriticalFailure);
+    }
+
+    #region Commit
+
+    private readonly Channel<Task<ConsumeResultAndProcessedMessageStatus>> _processedMessages;
+    private readonly Timer _timer;
+    private readonly object _commitLock = new();
+    private ConsumeResultAndProcessedMessageStatus? _lastConsumeResultAndProcessedMessageStatus;
+
+    private async Task ExecuteCommitLoopAsync(CancellationToken cancellationToken)
+    {
+        RestartCommitTimer();
+
+        while (!cancellationToken.IsCancellationRequested)
+        {
+            try
+            {
+                var result = await PeekAndAwaitProcessedMessages(cancellationToken);
+
+                if (IsIrrecoverableFailure(result.ProcessedMessageStatus))
+                {
+                    _applicationLifetime.StopApplication();
+                    break;
+                }
+
+                // Remove message from channel, when Task is successfully completed
+                await _processedMessages.Reader.ReadAsync(cancellationToken);
+
+                lock (_commitLock)
+                {
+                    _lastConsumeResultAndProcessedMessageStatus = result;
+                }
+
+                if (result.ConsumeResult.Offset.Value % _options.CommitPeriod == 0)
+                {
+                    Commit();
+                    RestartCommitTimer();
+                }
+            }
+            catch (Exception e) when (e is OperationCanceledException or ChannelClosedException)
+            {
+                break;
+            }
+        }
+
+        StopCommitTimer();
+    }
+
+    private async Task<ConsumeResultAndProcessedMessageStatus> PeekAndAwaitProcessedMessages(CancellationToken cancellationToken)
+    {
+        await _processedMessages.Reader.WaitToReadAsync(cancellationToken);
+
+        if (!_processedMessages.Reader.TryPeek(out var consumeAndProcessTask))
+        {
+            throw new InvalidOperationException("Awaited channel data has been removed by another consumer");
+        }
+
+        return await consumeAndProcessTask;
+    }
+
+    private void Commit()
+    {
+        lock (_commitLock)
+        {
+            if (_lastConsumeResultAndProcessedMessageStatus == null)
+            {
+                return;
+            }
+
+            try
+            {
+                _consumer?.Commit(_lastConsumeResultAndProcessedMessageStatus.ConsumeResult);
+                _lastConsumeResultAndProcessedMessageStatus = null;
+            }
+            catch (KafkaException e)
+            {
+                _logger.LogError(LogEvents.CommitError, e, "Commit error: {Reason}", e.Error.Reason);
+            }
+        }
+    }
+
+    private void RestartCommitTimer()
+    {
+        var autoCommitIntervalMs = _options.AutoCommitIntervalMs;
+        if (autoCommitIntervalMs != null)
+        {
+            _timer.Change(autoCommitIntervalMs.Value, Timeout.Infinite);
```

Comment:
I am surprised to learn that it is discouraged in most cases to use the built-in `period` of `Timer`. Is that to avoid issues if the callback takes longer than the configured `period`?

Reply:
I haven't read about that. I wanted to control the next commit myself. No idea, if that's the best design.